### PR TITLE
Add CNB Service Bindings to Image and Build resources

### PR DESCRIPTION
**Pull request by scothis**
The CNB Service Bindings spec provides a structured way to expose
service metadata and credentials into a build. For example, the build
can use the binding information to install a database driver, configure
an APM agent for the built image, or discover any other capability
outside of the built container.

Each binding reference has a name, a reference to a k8s ConfigMap
holding service metadata, and optionally a k8s Secret holding service
credentials. Builds generally should not have access to the credentials
so that the values are not backed into the built image, but instead
bound fresh at runtime.

For images, the bindings are located at `.spec.build.bindings`:

```
kind: Image
...
spec:
  ...
  build:
    bindings:
    - name: my-binding
      metadataRef:
        name: provider-binding-metadata
      secretRef:
        name: provider-binding-secret
```

For builds, the bindings are located at `.spec.bindings`:

```
kind: Build
...
spec:
  ...
  bindings:
  - name: my-binding
    metadataRef:
      name: provider-binding-metadata
    secretRef:
      name: provider-binding-secret
```

In the build the bindings appear as a directory tree under the path
specified by the `$CNB_BINDINGS` env var. Each binding is a directory,
which in turn has a `metadata` directory with files and if credentials
are exposed a `secret` directory with more files. The environment
variable and volume mounts are available during the detect and build
steps. Currently, the value of `$CNB_BINDINGS` is `/var/bindings` but
should not be presumed to be stable or consistent with the value at
runtime.

There is no guarantee the same services, or types of services, will be
bound at runtime.

Refs #369

### CODE REVIEW

File: pkg/apis/build/v1alpha1/build_pod.go [hunk 1]
```diff
@@ -84,6 +84,10 @@ var (
 		MountPath: "/builderPullSecrets",
 		ReadOnly:  true,
 	}
+	bindingsEnv = corev1.EnvVar{
+		Name:  "CNB_BINDINGS",
+		Value: "/var/bindings",
```
Comment (matthewmcnew) [inline pkg/apis/build/v1alpha1/build_pod.go:89]: It seems that [buildpacks spec](https://github.com/buildpacks/spec/blob/master/extensions/bindings.md#detect-and-build-phases) expects that this in `/platform/bindings`.

**Comment by matthewmcnew** [comment]
This is so good @scothis!  

I'd like to take to a moment to think through the security implications of allowing arbitrary secrets to be mounted in the build pod as you aptly noted in https://github.com/pivotal/kpack/issues/369.

File: pkg/apis/build/v1alpha1/build_pod.go (hunk 1 above)
Comment (scothis) [reply pkg/apis/build/v1alpha1/build_pod.go:89]: Good catch. I was paying attention to the launch phase which is variable based.

Comment (scothis) [reply pkg/apis/build/v1alpha1/build_pod.go]: Moved the mount path to be under `/platform/bindings`, I need to test it to make sure the nested volume mounts don't cause issues

**Comment by scothis** [comment]
Used the validator to disable the secretRef on the binding until we have a better understanding of how to protect again secret enumeration attacks, or direct capturing of secrets if they know the name.

**Comment by scothis** [comment]
Based on our conversation yesterday, I switch up the secret handling to forbid binding secrets that are attached to any service account in the namespace. This should offer a reasonable compromise to protect registry, and api server credentials while still allowing bindings to work as expected.

File: pkg/buildpod/generator_test.go [hunk 2]
```diff
@@ -184,6 +185,31 @@ func testGenerator(t *testing.T, when spec.G, it spec.S) {
...
+				K8sClient:       fakeK8sClient,
+				KeychainFactory: keychainFactory,
+				ImageFetcher:    imageFetcher,
+			}
+
+			var build = &testBuildPodable{
+				namespace: namespace,
+				bindings: []v1alpha1.Binding{
+					{
+						Name:        "naughty",
```
Comment (ashwin-venkatesh) [inline pkg/buildpod/generator_test.go:202]: lolol

File: samples/image_with_service_bindings.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
+apiVersion: build.pivotal.io/v1alpha1
```
Comment (ashwin-venkatesh) [inline samples/image_with_service_bindings.yaml:1]: Love the samples!!!

File: pkg/apis/build/v1alpha1/build_pod.go [hunk 4]
```diff
@@ -110,6 +110,51 @@ func (b *Build) BuildPod(config BuildPodImages, secrets []corev1.Secret, bc Buil
 		SubPath:   b.Spec.Source.SubPath, // empty string is a nop
 	}
 
+	bindingVolumes := []corev1.Volume{}
```
Comment (ashwin-venkatesh) [inline pkg/apis/build/v1alpha1/build_pod.go:113]: can this be moved into a private method `setupBindings()`?

Comment (scothis) [reply pkg/apis/build/v1alpha1/build_pod.go]: done

**Pull request merged**